File: sample-chat-obj-c/sample-chat/Components/ChatScreen/Views/ChatCollectionView/ChatCollectionViewFlowLayout/ChatCollectionViewFlowLayout.m
//
//  ChatCollectionViewFlowLayout.m
//  samplechat
//
//  Created by Injoit on 2/25/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

#import "ChatCollectionViewFlowLayout.h"
#import "ChatCollectionView.h"
#import "CollectionViewFlowLayoutInvalidationContext.h"
#import "ChatCellLayoutAttributes.h"

@interface ChatCollectionViewFlowLayout()
@property (strong, nonatomic) NSMutableDictionary *cache;
@end

@implementation ChatCollectionViewFlowLayout

@dynamic chatCollectionView;

- (ChatCollectionView *)chatCollectionView {
    return (id)self.collectionView;
}

//MARK: - Initialization

- (void)configureFlowLayout {
    
    self.scrollDirection = UICollectionViewScrollDirectionVertical;
    self.sectionInset = UIEdgeInsetsMake(10.0f, 4.0f, 10.0f, 4.0f);
    self.minimumLineSpacing = 4.0f;
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(didReceiveApplicationMemoryWarningNotification:)
                                                 name:UIApplicationDidReceiveMemoryWarningNotification
                                               object:nil];
    /**
     *  Init cache
     */
    self.cache = [[NSMutableDictionary alloc] init];
}

- (instancetype)init {
    self = [super init];
    if (self) {
        [self configureFlowLayout];
    }
    
    return self;
}

- (void)awakeFromNib {
    [super awakeFromNib];
    
    [self configureFlowLayout];
}

+ (Class)layoutAttributesClass {
    return [ChatCellLayoutAttributes class];
}

+ (Class)invalidationContextClass {
    return [CollectionViewFlowLayoutInvalidationContext class];
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (CGFloat)itemWidth {
    return CGRectGetWidth(self.collectionView.frame) - self.sectionInset.left - self.sectionInset.right;
}

//MARK: - Notifications

- (void)didReceiveApplicationMemoryWarningNotification:(NSNotification *)notification {
    [self resetLayout];
}

//MARK: - Collection view flow layout

- (UICollectionViewLayoutInvalidationContext *)invalidationContextForBoundsChange:(CGRect)newBounds {
    
    CollectionViewFlowLayoutInvalidationContext *context = [CollectionViewFlowLayoutInvalidationContext context];
    
    if (self.collectionView.bounds.size.width != newBounds.size.width) {
        context.invalidateFlowLayoutMessagesCache = YES;
    }
    
    return context;
}

- (void)invalidateLayoutWithContext:(CollectionViewFlowLayoutInvalidationContext *)context {
    
    if (context.invalidateDataSourceCounts) {
        context.invalidateFlowLayoutAttributes = YES;
        context.invalidateFlowLayoutDelegateMetrics = YES;
    }
    
    if (context.invalidateFlowLayoutMessagesCache) {
        
        [self resetLayout];
    }
    
    [super invalidateLayoutWithContext:context];
}

- (NSArray *)layoutAttributesForElementsInRect:(CGRect)rect {
    
    NSArray *attributesInRect = [[NSArray alloc] initWithArray:[super layoutAttributesForElementsInRect:rect]
                                                     copyItems:YES];
    
    [attributesInRect enumerateObjectsUsingBlock:^(ChatCellLayoutAttributes *attributesItem, NSUInteger idx, BOOL *stop) {
        
        if (attributesItem.representedElementCategory == UICollectionElementCategoryCell) {
            [self configureCellLayoutAttributes:attributesItem];
        }
    }];
    
    return attributesInRect;
}

- (UICollectionViewLayoutAttributes *)layoutAttributesForItemAtIndexPath:(NSIndexPath *)indexPath  {
    
    ChatCellLayoutAttributes *customAttributes = (id)[super layoutAttributesForItemAtIndexPath:indexPath];
    
    if (customAttributes.representedElementCategory == UICollectionElementCategoryCell) {
        [self configureCellLayoutAttributes:customAttributes];
    }
    
    return customAttributes;
}

- (BOOL)shouldInvalidateLayoutForBoundsChange:(CGRect)newBounds {
    
    CGRect oldBounds = self.collectionView.bounds;
    return !CGSizeEqualToSize(oldBounds.size, newBounds.size) ? YES : NO;
}

- (void)prepareForCollectionViewUpdates:(NSArray *)updateItems {
    
    [super prepareForCollectionViewUpdates:updateItems];
    
    [updateItems enumerateObjectsUsingBlock:^(UICollectionViewUpdateItem *updateItem, NSUInteger index, BOOL *stop) {
        
        if (updateItem.updateAction == UICollectionUpdateActionInsert) {
            
            CGFloat collectionViewHeight = CGRectGetHeight(self.collectionView.bounds);
            
            ChatCellLayoutAttributes *attributes =
            [ChatCellLayoutAttributes layoutAttributesForCellWithIndexPath:updateItem.indexPathAfterUpdate];
            
            if (attributes.representedElementCategory == UICollectionElementCategoryCell) {
                [self configureCellLayoutAttributes:attributes];
            }
            
            attributes.frame = CGRectMake(0.0f,
                                          collectionViewHeight + CGRectGetHeight(attributes.frame),
                                          CGRectGetWidth(attributes.frame),
                                          CGRectGetHeight(attributes.frame));
        }
    }];
}

//MARK:- Invalidation utilities
- (void)resetLayout {
    [self.cache removeAllObjects];
}

//MARK: - Message cell layout utilities
- (void)removeSizeFromCacheForItemID:(NSString *)itemID {
    [self.cache removeObjectForKey:itemID];
}

- (CGSize)containerViewSizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    
    //Item unique id
    NSString *itemID = [self.chatCollectionView.dataSource collectionView:self.chatCollectionView
                                                        itemIdAtIndexPath:indexPath];
    
    NSValue *cachedSize = [self.cache objectForKey:itemID];
    
    if (cachedSize != nil) {
        
        return [cachedSize CGSizeValue];
    }
    
    ChatCellLayoutModel layoutModel =
    [self.chatCollectionView.delegate collectionView:self.chatCollectionView
                              layoutModelAtIndexPath:indexPath];
    
    CGSize finalSize;
    
    if (CGSizeEqualToSize(layoutModel.staticContainerSize, CGSizeZero)) {
        
        //  from the cell xibs, there is a 2 point space between avatar and bubble
        CGFloat spacingBetweenAvatarAndBubble = 2.0f;
        CGFloat horizontalContainerInsets = layoutModel.containerInsets.left + layoutModel.containerInsets.right;
        CGFloat horizontalInsetsTotal = horizontalContainerInsets + spacingBetweenAvatarAndBubble;
        CGFloat maximumWidth = self.itemWidth - layoutModel.avatarSize.width - layoutModel.maxWidthMarginSpace;
        
        if (layoutModel.maxWidth > 0) {
            maximumWidth = MIN(maximumWidth, layoutModel.maxWidth - layoutModel.avatarSize.width - layoutModel.maxWidthMarginSpace);
        }
        NSAssert(maximumWidth >= 0, @"Maximum width cannot be a negative nuber. Please check your maxWidthMarginSpace value.");
        
        CGSize dynamicSize = [self.chatCollectionView.delegate collectionView:self.chatCollectionView
                                                       dynamicSizeAtIndexPath:indexPath
                                                                     maxWidth:maximumWidth - horizontalInsetsTotal];
        CGFloat verticalContainerInsets =
        layoutModel.containerInsets.top + layoutModel.containerInsets.bottom + layoutModel.topLabelHeight + layoutModel.bottomLabelHeight;
        
        CGFloat additionalSpace =
        layoutModel.spaceBetweenTextViewAndBottomLabel + layoutModel.spaceBetweenTopLabelAndTextView;
        
        CGFloat finalWidth = dynamicSize.width + horizontalContainerInsets;
        
        CGFloat cellHeight = dynamicSize.height + verticalContainerInsets + additionalSpace;
        CGFloat finalCellHeight = MAX(cellHeight, layoutModel.avatarSize.height);
        
        CGFloat minWidth = [self.chatCollectionView.delegate collectionView:self.chatCollectionView
                                                        minWidthAtIndexPath:indexPath];
        minWidth += horizontalContainerInsets;
        
        finalSize = CGSizeMake(MIN(MAX(finalWidth, minWidth), maximumWidth), finalCellHeight);
    }
    else {
        
        finalSize = layoutModel.staticContainerSize;
    }
    
    [self.cache setObject:[NSValue valueWithCGSize:finalSize] forKey:itemID];
    
    return finalSize;
}

- (CGSize)sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    CGSize containerSize = [self containerViewSizeForItemAtIndexPath:indexPath];
    return CGSizeMake(self.itemWidth, ceilf(containerSize.height));
}

- (void)configureCellLayoutAttributes:(ChatCellLayoutAttributes *)layoutAttributes {
    
    NSIndexPath *indexPath = layoutAttributes.indexPath;
    
    CGSize containerSize = [self containerViewSizeForItemAtIndexPath:indexPath];
    layoutAttributes.containerSize = containerSize;
    
    // fix for content size changes (example: split view display mode change)
    CGRect frame = layoutAttributes.frame;
    frame.origin.x = self.sectionInset.left;
    frame.size.width = [self itemWidth];
    layoutAttributes.frame = frame;
    
    ChatCellLayoutModel layoutModel =
    [self.chatCollectionView.delegate collectionView:self.chatCollectionView
                              layoutModelAtIndexPath:indexPath];
    
    layoutAttributes.avatarSize = layoutModel.avatarSize;
    layoutAttributes.containerInsets = layoutModel.containerInsets;
    layoutAttributes.topLabelHeight = layoutModel.topLabelHeight;
    layoutAttributes.bottomLabelHeight = layoutModel.bottomLabelHeight;
    layoutAttributes.spaceBetweenTopLabelAndTextView = layoutModel.spaceBetweenTopLabelAndTextView;
    layoutAttributes.spaceBetweenTextViewAndBottomLabel = layoutModel.spaceBetweenTextViewAndBottomLabel;
}

@end

File: sample-chat-swift/sample-chat-swift/Components/ChatScreen/Chat/CollectionView/ChatCollectionViewFlowLayout.swift
//
//  ChatCollectionViewFlowLayout.swift
//  sample-chat-swift
//
//  Created by Injoit on 1/28/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

import UIKit
/**
 *  The `ChatCollectionViewFlowLayout` is a concrete layout object that inherits
 *  from `UICollectionViewFlowLayout` and organizes message items in a vertical list.
 *  Each `ChatCollectionViewCell` in the layout can display messages of arbitrary sizes and avatar images,
 *  as well as metadata such as a timestamp and sender.
 *  You can easily customize the layout via its properties or its delegate methods
 *  defined in `ChatCollectionViewDelegateFlowLayout`.
 *
 *  @see ChatCollectionViewDelegateFlowLayout.
 *  @see ChatCollectionViewCell.
 */
class ChatCollectionViewFlowLayout: UICollectionViewFlowLayout {
    
    /**
     *  The collection view object currently using this layout object.
     */
    lazy private(set) var chatCollectionView: ChatCollectionView = {
        return collectionView as! ChatCollectionView
    }()
    /**
     *  Returns the width of items in the layout.
     */
    
    var itemWidth: CGFloat {
        guard let collectionView = collectionView else {
            return 0.0
        }
        return collectionView.frame.width - sectionInset.left - sectionInset.right
    }
    /**
     *  Size for item and index path.
     *
     *  @discussion Returns cached size of item. If size is not in cache, then counts it first.
     *
     *  @return Size of item at index path
     */
    
    func sizeForItem(at indexPath: IndexPath?) -> CGSize {
        
        var height = containerViewSizeForItem(at: indexPath).height
        height.round(.up)
        return CGSize(width: itemWidth, height: height)
    }
    /**
     *  Removing size for itemID from cache.
     *
     *  @discussion Use this method before any of collection view reload methods, that will update missed cached sizes.
     */
    //MARK: - Message cell layout utilities
    func removeSizeFromCache(forItemID itemID: String?) {
        cache.removeValue(forKey: itemID)
    }
    
    
    private var cache: [AnyHashable : Any] = [:]
    
    //MARK: - Initialization
    
    func configureFlowLayout() {
        
        scrollDirection = UICollectionView.ScrollDirection.vertical
        sectionInset = UIEdgeInsets(top: 10.0, left: 0.0, bottom: 10.0, right: 0.0)
        minimumLineSpacing = 16.0
        
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(didReceiveApplicationMemoryWarning(_:)),
                                               name: UIApplication.didReceiveMemoryWarningNotification,
                                               object: nil)
        /**
         *  Init cache
         */
        cache = [AnyHashable : Any]()
    }
    
    override init() {
        super.init()
        
        configureFlowLayout()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    override func awakeFromNib() {
        
        super.awakeFromNib()
        configureFlowLayout()
    }
    
    override class var layoutAttributesClass: AnyClass {
        return ChatCellLayoutAttributes.self
    }
    
    override class var invalidationContextClass: AnyClass {
        return CollectionViewFlowLayoutInvalidationContext.self
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
    }
    
    //MARK: - Notifications
    @objc func didReceiveApplicationMemoryWarning(_ notification: Notification?) {
        resetLayout()
    }
    
    //MARK: - Collection view flow layout
    override func invalidationContext(forBoundsChange newBounds: CGRect) -> UICollectionViewLayoutInvalidationContext {
        
        let context = CollectionViewFlowLayoutInvalidationContext()
        if let collectionView = collectionView, collectionView.bounds.size.width != newBounds.size.width {
            context.invalidateFlowLayoutMessagesCache = true
        }
        return context
    }
    
    override func invalidateLayout(with context: UICollectionViewLayoutInvalidationContext) {
        guard let invalidationContext = context as? CollectionViewFlowLayoutInvalidationContext else {
            super.invalidateLayout(with: context)
            return
        }
        if invalidationContext.invalidateDataSourceCounts == true {
            invalidationContext.invalidateFlowLayoutAttributes = true
            invalidationContext.invalidateFlowLayoutDelegateMetrics = true
        }
        
        if invalidationContext.invalidateFlowLayoutMessagesCache == false  {
            resetLayout()
        }
        super.invalidateLayout(with: context)
    }
    
    override func layoutAttributesForElements(in rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
        let attributesInRect = super.layoutAttributesForElements(in: rect)
        attributesInRect?.forEach {
            if $0.representedElementCategory == .cell, let item = $0 as? ChatCellLayoutAttributes {
                configureCellLayoutAttributes(item)
            }
        }
        return attributesInRect
    }

    override func layoutAttributesForItem(at indexPath: IndexPath) -> UICollectionViewLayoutAttributes? {
        
        let customAttributes = super.layoutAttributesForItem(at: indexPath) as? ChatCellLayoutAttributes
        
        if customAttributes?.representedElementCategory == .cell {
            configureCellLayoutAttributes(customAttributes)
        }
        
        return customAttributes
    }
    
    override func shouldInvalidateLayout(forBoundsChange newBounds: CGRect) -> Bool {
        let oldBounds = collectionView?.bounds ?? .zero
        return oldBounds.size.equalTo(newBounds.size) == false
    }
    
    override func prepare(forCollectionViewUpdates updateItems: [UICollectionViewUpdateItem]) {
        super.prepare(forCollectionViewUpdates: updateItems)
        
        guard let collectionView = collectionView  else {
            return
        }
        
        let collectionHeight = collectionView.bounds.height
        
        updateItems.forEach{
            if $0.updateAction == .insert, let index = $0.indexPathAfterUpdate {
                let attributes = ChatCellLayoutAttributes(forCellWith:index)
                if attributes.representedElementCategory == .cell {
                    configureCellLayoutAttributes(attributes)
                }
                attributes.frame = CGRect(x: 0.0,
                                          y: collectionHeight + attributes.frame.height,
                                          width: attributes.frame.width,
                                          height: attributes.frame.height)
            }
        }
    }
    
    //MARK:- Invalidation utilities
    func resetLayout() {
        cache.removeAll()
    }
    
    private func containerViewSizeForItem(at indexPath: IndexPath?) -> CGSize {
        
        guard let indexPath = indexPath else {
            return .zero
        }
        
        guard let chatDataSource = chatCollectionView.chatDataSource else {
            return .zero
        }
        
        let itemID = chatDataSource.collectionView(chatCollectionView, itemIdAt: indexPath)
        
        if let cachedSize = cache[itemID] as? NSValue {
            return cachedSize.cgSizeValue
        }
        
        guard let chatDelegate = chatCollectionView.chatDelegate else {
            return .zero
        }
        
        let layoutModel = chatDelegate.collectionView(chatCollectionView, layoutModelAt: indexPath)
        
        var finalSize = layoutModel.staticContainerSize
        
        if layoutModel.staticContainerSize.equalTo(.zero) {
            
            //  from the cell xibs, there is a 2 point space between avatar and bubble
            let spacingBetweenAvatarAndBubble: CGFloat = 16.0
            let horizontalContainerInsets = layoutModel.containerInsets.left + layoutModel.containerInsets.right
            let horizontalInsetsTotal: CGFloat = horizontalContainerInsets + spacingBetweenAvatarAndBubble
            var maximumWidth: CGFloat = itemWidth - layoutModel.avatarSize.width - layoutModel.maxWidthMarginSpace
            
            if layoutModel.maxWidth > 0.0 {
                maximumWidth = min(maximumWidth, layoutModel.maxWidth - layoutModel.avatarSize.width - layoutModel.maxWidthMarginSpace)
            }
            assert(maximumWidth >= 0.0, "Maximum width cannot be a negative nuber. Please check your maxWidthMarginSpace value.")
            
            let dynamicSize = chatDelegate.collectionView(chatCollectionView, dynamicSizeAt: indexPath, maxWidth: maximumWidth - horizontalInsetsTotal)

            let verticalContainerInsets = layoutModel.containerInsets.top + layoutModel.containerInsets.bottom + layoutModel.timeLabelHeight + 6.0
            
            let additionalSpace = layoutModel.spaceBetweenTopLabelAndTextView
            
            let finalWidth = dynamicSize.width + horizontalContainerInsets
            
            let cellHeight = dynamicSize.height + verticalContainerInsets + additionalSpace
            
            let finalCellHeight = max(cellHeight, layoutModel.avatarSize.height)
            
            var minWidth = chatDelegate.collectionView(chatCollectionView, minWidthAt: indexPath)
            minWidth += horizontalContainerInsets
            
            finalSize = CGSize(width: min(max(finalWidth, minWidth), maximumWidth), height: finalCellHeight)
        }
        
        cache[itemID] = NSValue(cgSize: finalSize)
        
        return finalSize
    }
    
    func configureCellLayoutAttributes(_ layoutAttributes: ChatCellLayoutAttributes?) {
        
        guard let layoutAttributes = layoutAttributes else {
            return
        }
        
        let containerSize = containerViewSizeForItem(at: layoutAttributes.indexPath)
        layoutAttributes.containerSize = containerSize
        
        // fix for content size changes (example: split view display mode change)
        var frame = layoutAttributes.frame
        frame.origin.x = sectionInset.left
        frame.size.width = itemWidth
        layoutAttributes.frame = frame
        
        guard let chatDelegate = chatCollectionView.chatDelegate else {
            return
        }
        
        let layoutModel = chatDelegate.collectionView(chatCollectionView, layoutModelAt: layoutAttributes.indexPath)
        
        layoutAttributes.avatarSize = layoutModel.avatarSize
        layoutAttributes.containerInsets = layoutModel.containerInsets
        layoutAttributes.topLabelHeight = layoutModel.timeLabelHeight
        layoutAttributes.spaceBetweenTopLabelAndTextView = layoutModel.spaceBetweenTopLabelAndTextView
    }
}
